Handle special case for PolarManifold::normal_vector

gassmoeller opened the pull request:
This copies a special case from `SphericalManifold::normal_vector` to `PolarManifold`. The issue at hand is that for horizontal boundaries in spheres (i.e. r=const.) and when using a Q1 mapping the radius of the point `p` that is handed over to the function can differ significantly from the radius at the cell vertices that are used to compute the normal vector in `Manifold::normal_vector`. This in particular leads to issues when computing normal flux constraints, because the normal vector is not normal to the radial surface (see geodynamics/aspect#3359 for a bug report related to this behavior). For comparison this is some of the output of the included testcase without the fix (the normal from spherical manifold and polar manifold should be identical):
```
DEAL::Position: 0.559808 0.150000
DEAL::Polar normal: 0.922712 0.385489
DEAL::Spherical normal: 0.965926 0.258819
DEAL::Relative difference: 0.133838
DEAL::
DEAL::Position: 0.279904 0.0750000
DEAL::Polar normal: 0.922712 0.385489
DEAL::Spherical normal: 0.965926 0.258819
DEAL::Relative difference: 0.133838
```

drwells commented:
Since both of these methods are defined in the same file could we just move this function into a free function defined in an anonymous namespace?

gassmoeller commented:
something like the following?
```
namespace
{
  bool horizontal_spherical_face(face, center)
    {
      ...
    }
}

PolarManifold::normal_vector(...)
{
  if (horizontal_spherical_face(face,center))
    {
            const Tensor<1, spacedim> unnormalized_spherical_normal = p - center;
      const Tensor<1, spacedim> normalized_spherical_normal =
        unnormalized_spherical_normal / unnormalized_spherical_normal.norm();
      return normalized_spherical_normal;
    }

  return Manifold::normal_vector(...)
}
```
Sure that would work. Do you want me to change it?

drwells commented:
Yup :)

drwells commented:
Yup :)

gassmoeller commented:
I refactored the code and incorporated the changes of #9559. Also fixed a warning about reaching the end of a non-void function without return I got in gcc 9.2.1 (likely introduced in #9559).

bangerth commented on include/deal.II/grid/manifold_lib.h:
```diff
@@ -110,6 +110,14 @@ class PolarManifold : public ChartManifold<dim, spacedim, spacedim>
   push_forward_gradient(const Point<spacedim> &chart_point) const override;
 
+  /**
+   * Return the (normalized) normal vector at the point @p p.
```
Can you just `@copydoc` from the base class (or actually just say `// documentation is copied from the base class`

bangerth replied:
This won't work -- if you use `/** ... */` then that's what doxygen is going to take. If you want doxygen to just use the base class function's documentation, then you need to use a non-doxygen comment like `// ...` here.

bangerth commented on source/grid/manifold_lib.cc:
```diff
@@ -462,6 +508,8 @@ SphericalManifold<dim, spacedim>::normal_vector(
     // base class.
     return Manifold<dim, spacedim>::normal_vector(face, p);
+
+  return Tensor<1,spacedim>();
```
Why would this be necessary? You terminate both the `if` and `else` branch above in a `return` statement...

drwells replied:
I think this was to fix a warning

gassmoeller replied:
Yes, without this I get a warning about the possibility of never reaching a return statement in gcc 9.2.1. Not sure why, `if-else` should always reach a return, but I guess gcc is worried that one branch may contain a `break` statement before the `return`. Anyway, this is something that came up in a few places in ASPECT already (and likely in deal.II in other places too) and our usual solution was to just add a return that will never trigger.

bangerth commented:
@luca-heltai -- do you agree that this looks correct?

bangerth commented:
It looks like you need to reindent, though.

gassmoeller commented:
I modified the documentation and reindented. Should be ready to go.

bangerth commented on source/grid/manifold_lib.cc:
```diff
@@ -274,6 +274,77 @@ PolarManifold<dim, spacedim>::push_forward_gradient(
+    // base class.
+    return Manifold<dim, spacedim>::normal_vector(face, p);
+
+  return Tensor<1, spacedim>();
```
Hm, how bizarre. But let's keep this then.

gassmoeller commented:
> This won't work -- if you use /** ... */ then that's what doxygen is going to take. If you want doxygen to just use the base class function's documentation, then you need to use a non-doxygen comment like // ... here.

Is that the general convention for how to handle duplicated doxygen documentation? Just asking, because I have not seen it anywhere else, and `SphericalManifold::normal_vector` has the same documentation (also copied from the base class).

bangerth commented:
I think we took a few years to learn about `@copydoc`, and so it's not so widely used. Here's an example from `manifold_lib.h`:
```
  /**
   * @copydoc ChartManifold::pull_back()
   */
  virtual Point<spacedim>
  pull_back(const Point<spacedim> &space_point) const override;
```
But even that is strictly speaking unnecessary because doxygen will, if you don't document a virtual function in a derived class, simply copy the documentation from the base class. Like here in `mapping_fe_field.h`:
```
  // for documentation, see the Mapping base class
  virtual void
  transform(const ArrayView<const Tensor<1, dim>> &                  input,
            const MappingKind                                        kind,
            const typename Mapping<dim, spacedim>::InternalDataBase &internal,
            const ArrayView<Tensor<1, spacedim>> &output) const override;
```
which shows up here: https://dealii.org/developer/doxygen/deal.II/classMappingFEField.html#a888203d8543663a7ab76e750430798c9

gassmoeller commented:
I updated it. I knew about `@copydoc`, but I had never seen the `// ...` alternative, and I was not sure which version you wanted me to change to.

bangerth commented:
/rebuild
